### PR TITLE
async_unary_call: add a Destroy method to ClientAsyncResponseReaderInterface

### PR DESCRIPTION
**Pull request by jacobsa**
Before this commit, we specialized `std::default_delete` for
`ClientAsyncResponseReaderInterface` to do nothing. This means that, despite the
fact that async stub methods return
`std::unique_ptr<ClientAsyncResponseReaderInterface>`, it's impossible to have a
subclass of that interface that is allocated on the heap or whose destructor has
side effects without a hack that manages ownership _outside of_ the `unique_ptr`
returned by the stub.

This was introduced by dd36b153, which introduced the specialization for
`grpc::ClientAsyncResponseReader` itself, and e8a61d63b, which did so
for the interface. Neither of them left any explanation. :-/

I think the intent is a combination of the following facts:

*   In production uses, `ClientAsyncResponseReaderHelper::Create` allocates the
    `ClientAsyncResponseReader` on an arena.

*   Although it's technically non-trivial, the `ClientAsyncResponseReader`
    destructor doesn't actually have any important side effects.

*   Therefore it's not necessary to run the destructor, and it's not correct to
    use `delete` to free the backing memory.

From this point of view, the original commit (`dd36b153`) is technically
correct: `std::unique_ptr<ClientAsyncResponseReader>` doesn't actually need to
do anything. But:

*   Of course the objects may be deleted other ways. If I have a
    `std::unique_ptr<ClientAsyncResponseReaderInterface>` (i.e. without a
    specific deleter type), it would be natural for me to assume that I can do
    `delete p.release()` to get the same effect as destroying the unique
    pointer. But that is not true, and doing so will be undefined behavior.

    So from this point of view `dd36b153` is not correct, or at least makes for
    a surprising API.

*   `ClientAsyncResponseReader` is not the only subclass of the interface. The
    interface's destructor is virtual, so other subclasses may have destructors
    with side effects. However they are never actually run! This is a huge
    surprise: it took me hours yesterday to figure out that I wasn't actually
    crazy: `delete p.release()` really was working differently than just letting
    a `std::unique_ptr p` go out of scope.

    So from this point of view `e8a61d63` is also not correct.

This is not a theoretical concern: I've found multiple hacks in user
code working around this with a grumbling tone. For example, see in the
cloud bigtable library that uses the phrase "terrible, horrible, no
good, very bad hack":

    https://github.com/googleapis/google-cloud-cpp/blob/74de20227371119ea8b8c788d37ae4e24be8ed5d/google/cloud/bigtable/testing/mock_response_reader.h#L77-L106

This commit makes it possible for subclasses to opt into reasonable behavior by
defining an override for `Destroy` that calls `delete this`. We change the
specialization to call `Destroy` in order to make this possible. There is no
behavior change in existing code. I will follow up separately with a change to
the base implementation of `Destroy`, making the reasonable behavior opt-out
instead`, which will be a behavior change for any subclass that depends on the
surprising default.

@soheilhy @vjpai

### CODE REVIEW

**Review by soheilhy — APPROVED** [review]
Thank you for the change! Looks great.

I have added @markdroth, @ctiller, @vjpai and @yashykt. I will merge it as soon as we get one more LGTM.

**Review by vjpai — CHANGES_REQUESTED** [review]
We have never claimed to support mocking of async calls; any use of that is inherently a hack. Also, this is an API addition, so it needs a GRFC.. That would need to marinate for two weeks, in which case my review doesn't really mean much. ( Or maybe rename it to InternalDestroy and mark very clearly that even though it is a virtual function, it is not API.) So, I'll leave it to y'all as to how to move forward. If it's the GRFC route, I can remove myself from this review.

**Comment by ctiller** [comment]
I think if we expose an interface it's a reasonable expectation that someone will want to implement it however they see fit, and that we should support freeing that memory.

That said, yeah we'll need to do the gRFC thing for API changes... happy to help guide this through.

**Comment by vjpai** [comment]
Ok, if we're headed to gRFC, I'm going to remove myself from the review and dismiss my requested changes.

**Comment by jacobsa** [comment]
@ctiller: thanks, that was my reasoning too. The fact that there is a virtual destructor implies that the destructor should be able to do things. Can you tell me what exactly I need to do here? I'm not familiar with the process.

**Comment by jacobsa** [comment]
(By the way, this is the only reason you _can't_ support mocking async calls, I think. I've got a prototype of mocking working just fine with this change.)

**Comment by jacobsa** [comment]
@ctiller Nevermind, I found the instructions. Could you please have a look at [the proposal](https://github.com/grpc/proposal/pull/238)? Should I close this thread in the meantime?

**Comment by jacobsa** [comment]
Okay, I'm going to close this for now. I'll open a new one once [the proposal](https://github.com/grpc/proposal/pull/238) is approved. Sorry for the premature PR.

**Comment by vjpai** [comment]
You don't need to close it. In fact, I recommend keeping it open so that the tests have run and it can be merged as soon as the 14 day waiting period expires; just don't merge it until the gRFC lands.

**Comment by jacobsa** [comment]
Ah I see, sorry. It's out of date based on what's in the RFC though, so I'll let it die and send a new one. I'm currently testing it internally.

**Pull request closed without merging**
